rust/template/main.rs

```diff
@@ -28,74 +28,71 @@ use differential_datalog::program::*;
 use cmd_parser::*;
 use differential_datalog::record::*;
 use time::precise_time_ns;
+use api::HDDlog;
+use std::io::Write;
 
 // uncomment to enable profiling
 //extern crate cpuprofiler;
 //use cpuprofiler::PROFILER;
 
-extern "C" fn print_fun(_arg: libc::uintptr_t,
-                        table: libc::size_t,
-                        rec: *const Record,
-                        pol: bool) {
-    unsafe {
-        eprintln!("{} {:?} {}", if pol { "insert" } else { "delete" }, table, *rec);
-    }
-}
-
-fn handle_cmd(handler: &Box<dyn IMTUpdateHandler<Value>>,
-              db: &Arc<Mutex<ValMap>>,
-              deltadb: &Arc<Mutex<Option<DeltaMap>>>,
-              p: &mut RunningProgram<Value>,
-              upds: &mut Vec<Update<Value>>,
+fn handle_cmd(hddlog: &HDDlog,
+              print_deltas: bool,
+              upds: &mut Vec<UpdCmd>,
               cmd: Command) -> (i32, bool)
 {
     let resp = (
         if !is_upd_cmd(&cmd) {
-            apply_updates(p, upds)
+            apply_updates(hddlog, upds)
         } else {
             Ok(())
         }).and(match cmd {
         Command::Start => {
-            p.transaction_start()
+            hddlog.transaction_start()
         },
         Command::Commit(record_delta) => {
-            if record_delta {
-                *deltadb.lock().unwrap() = Some(DeltaMap::new());
-            }
             // uncomment to enable profiling
-            handler.before_commit();
             //PROFILER.lock().unwrap().start("./prof.profile").expect("Couldn't start profiling");
-            let res = p.transaction_commit();
+            let mut current_table = None;
+            let res = if record_delta {
+                let cb = if print_deltas {
+                    Some(move |table, val: &Record, pol: bool| {
```

Comment:
The code still compiles without `move`, which is surprising. What's going on here with `current_table`? Maybe Rust is automatically converting it into a mutable reference?

Reply:
I see, it doesn't have to be a reference. `FnMut` can capture any mutable variable